libraries/extractor/src/main/java/androidx/media3/extractor/text/ssa/SsaParser.java

```diff
@@ -461,11 +426,104 @@ private static Cue createCue(
       cue.setPosition(styleOverrides.position.x / screenWidth);
       cue.setLine(styleOverrides.position.y / screenHeight, LINE_TYPE_FRACTION);
     } else {
-      // TODO: Read the MarginL, MarginR and MarginV values from the Style & Dialogue lines.
       cue.setPosition(computeDefaultLineOrPosition(cue.getPositionAnchor()));
       cue.setLine(computeDefaultLineOrPosition(cue.getLineAnchor()), LINE_TYPE_FRACTION);
     }
 
+    // Apply margins if there are no overrides and we have valid positions.
+    if (styleOverrides.alignment == SsaStyle.SSA_ALIGNMENT_UNKNOWN
+        && styleOverrides.position == null
+        && cue.getPosition() != Cue.DIMEN_UNSET
+        && cue.getLine() != Cue.DIMEN_UNSET) {
+
+      // Margin from Dialogue lines takes precedence over margin from Style line.
+      float marginLeft = dialogueMarginLeft != 0f
+          ? dialogueMarginLeft / screenWidth
+          : style != null ? style.marginLeft / screenWidth : 0f;
+      float marginRight = dialogueMarginRight != 0f
+          ? dialogueMarginRight / screenWidth
+          : style != null ? style.marginRight / screenWidth : 0f;
+
+      // Apply margin left, margin right.
+      if (SsaStyle.hasLeftAlignment(style)) {
+        cue.setPosition(cue.getPosition() + marginLeft);
+        cue.setSize(1 - marginRight - marginLeft);
+      } else if (SsaStyle.hasRightAlignment(style)) {
+        cue.setPosition(cue.getPosition() - marginRight);
+        cue.setSize(1 - marginRight - marginLeft);
+      } else {
+        // Center alignment or unknown.
+        cue.setPosition(cue.getPosition() + (marginLeft - marginRight) / 2);
```

Comment:
I'm not sure I understand the maths here (or for the `hasRightAlignment` case above).

`Cue.position` always measures from the left hand side of the screen (for horizontal text cues), no matter what value `Cue.textAlignment` has. So shouldn't it always be increased by `marginLeft`?

----

In general applying these margins after the size/position has already been calculated seems hard to reason about - why not apply them as part of computing the size and position?